serialization/tests/derive.rs

```diff
@@ -0,0 +1,229 @@
+#![allow(clippy::unnecessary_cast)]
+
+use serialization::{tagged::*, *};
+
+use proptest::prelude::*;
+
+// A couple of types to experiment with
+
+#[derive(PartialEq, Eq, Clone, Tagged, Encode, Decode, Debug)]
+enum Enum0 {
+    This = 55,
+}
+
+#[derive(PartialEq, Eq, Clone, Tagged, Encode, Decode, Debug)]
+enum Enum1 {
+    #[codec(index = 42)]
+    The(u16, String),
+}
+
+#[derive(PartialEq, Eq, Clone, DirectEncode, DirectDecode, Debug)]
+enum Enum2 {
+    X(Enum0),
+    Y(Enum1),
+    Z { tag: Tag<7>, three_bytes: [u8; 3] },
+}
+
+#[derive(PartialEq, Eq, Tagged, Encode, Decode, Debug)]
+struct Struct0 {
+    ver: Tag<25>,
+    data: u64,
+    extra: Vec<u8>,
+}
+
+#[derive(PartialEq, Eq, Tagged, Encode, Decode, Debug)]
+struct Struct1 {
+    struct0: Struct0,
+    more_data: Option<[u8; 32]>,
+}
+
+#[derive(PartialEq, Eq, Tagged, Encode, Decode, Debug)]
+struct Struct2<T> {
+    nested: T,
+    stuff: u128,
+}
+
+#[derive(PartialEq, Eq, Tagged, Encode, Decode, Debug)]
+struct Struct3<const N: u8> {
+    version: Tag<N>,
+    int: u32,
+}
+
+// Generators for the types
+
+prop_compose! {
+    fn gen_enum1()(n: u16, s: String) -> Enum1 {
+        Enum1::The(n, s)
+    }
+}
+prop_compose! {
+    fn gen_struct0()(data: u64, extra: Vec<u8>) -> Struct0 {
+        Struct0 { ver: Default::default(), data, extra }
+    }
+}
+prop_compose! {
+    fn gen_struct1()(struct0 in gen_struct0(), more_data: Option<[u8; 32]>) -> Struct1 {
+        Struct1 { struct0, more_data }
+    }
+}
+prop_compose! {
+    fn gen_struct2_enum1()(nested in gen_enum1(), stuff: u128) -> Struct2<Enum1> {
+        Struct2 { nested, stuff }
+    }
+}
+prop_compose! {
+    fn gen_struct3_75()(int: u32) -> Struct3<75> {
+        Struct3 { version: Tag::default(), int }
+    }
+}
+fn gen_enum2() -> impl Strategy<Value = Enum2> {
+    prop_oneof![
+        Just(Enum2::X(Enum0::This)),
+        gen_enum1().prop_map(Enum2::Y),
+        any::<[u8; 3]>().prop_map(|three_bytes| Enum2::Z {
+            tag: Tag::default(),
+            three_bytes
+        }),
+    ]
+}
+
+// Test helpers
+
+fn check_initial_byte<T: Tagged + Encode>(x: &T) {
+    x.using_encoded(|encoded| {
+        assert_eq!(encoded[0], T::TAG, "tag mismatch");
+    });
+}
+
+fn check_codec<T: Eq + std::fmt::Debug + Encode + Decode>(original: &T) {
+    let encoded = original.encode();
+    let decoded = T::decode(&mut &encoded[..]).ok();
```

Comment:
In tests, I always prefer to use unwrap() instead of compared with Some() or None. The benefit is that when you unwrap, an error will print the error for you.

Reply:
I changed this to compare to `Ok`. That way the error is not lost and we get both the error and expected value printed out on failure. I am not sure why I converted to `Option` in the first place, probably some sort of error that I assumed would be solved by it so it's an oversight on my part.